bind/gen.go

```diff
@@ -203,9 +203,44 @@ func GoPyMainRun() {
 # File is generated by gopy. Do not edit.
 # %[2]s
 
-from pybindgen import retval, param, Module
+from pybindgen import retval, param, Function, Module
 import sys
 
+class CheckedFunction(Function):
+    def __init__(self, *a, **kw):
+        super(CheckedFunction, self).__init__(*a, **kw)
+        self._failure_expression = kw.get('failure_expression', '')
+        self._failure_cleanup = kw.get('failure_cleanup', '')
+
+    def set_failure_expression(self, expr):
+        self._failure_expression = expr
+
+    def set_failure_cleanup(self, expr):
+        self._failure_cleanup = expr
+
+    def generate_call(self):
+        super(CheckedFunction, self).generate_call()
+        check = "PyErr_Occurred()"
+        if self._failure_expression:
+            check = "{} && {}".format(self._failure_expression, check)
+        failure_cleanup = self._failure_cleanup or None
+        self.before_call.write_error_check(check, failure_cleanup)
+
+def add_checked_function(mod, name, retval, params, failure_expression='', *a, **kw):
+    fn = CheckedFunction(name, retval, params, *a, **kw)
+    fn.set_failure_expression(failure_expression)
+    mod._add_function_obj(fn)
+    return fn
+
+def add_checked_string_function(mod, name, retval, params, failure_expression='', *a, **kw):
+    fn = CheckedFunction(name, retval, params, *a, **kw)
+    cleanup = 'free(retval);'
+    fn.set_failure_cleanup(cleanup)
```

Comment:
This is doing a `free(retval)` when an error has been detected. But is it safe to assume we will always keep allocating an empty string even on exception, in the Go code? Should I change this to:
```C
if (retval != NULL) free(retval);
```

Reply:
commit 57a66d6